command/agent/agent.go

```diff
@@ -1038,19 +1064,28 @@ func (a *Agent) Shutdown() error {
 		// http server is HTTPS if TLSConfig is not nil and NextProtos does not only contain "h2"
 		// the latter seems to be a side effect of HTTP/2 support in go 1.8. TLSConfig != nil is
 		// no longer sufficient to check for an HTTPS server.
-		if srv.proto == "https" {
-			a.logger.Println("[INFO] agent: Stopping HTTPS server", srv.Addr)
-		} else {
-			a.logger.Println("[INFO] agent: Stopping HTTP server", srv.Addr)
-		}
+		a.logger.Printf("[INFO] agent: Stopping %s server %s",
+			strings.ToUpper(srv.proto), srv.Addr)
 
 		// old behavior: just die
 		// srv.Close()
 
 		// graceful shutdown
 		ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
 		defer cancel()
-		srv.Shutdown(ctx)
+
+		done := make(chan struct{})
+		go func() {
+			srv.Shutdown(ctx)
+			close(done)
+		}()
+		select {
+		case <-done:
+			// server down within timeout
+		case <-ctx.Done():
```

Comment:
yes, and a good idea as well since we're timing out every time. The server doesn't shut down cleanly and I need to figure out why. I've refactored it to not use the second channel and will investigate the code that hangs the server during shutdown.